### PR TITLE
Fix #89: Use ES6

### LINKED ISSUE #89: Use ES6/7
Use Babel to transpile it, and drop JQuery by the way if possible.

And I think I'll start working on it right now. :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by Bat41**
We now use ES6 for JavaScript. We use Babel to transpile it. I also removed JQuery and all the other scripts, so we only have "vanilla" JS.

### CODE REVIEW

File: data/scripts/valadoc.babel.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+          location = html.searchFocused.children[0].href
+        }
+      case 17: // ctrl
+        html.searchField.focus() // we focus the search
+        break
+    }
+  })
+})
+
+let searchDelay
```
Comment (btkostner) [inline data/scripts/valadoc.babel.js]: Thanks for adding a debounce. This was really annoying.

File: data/scripts/valadoc.babel.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
+// Some configuration that shouldn't be modified at runtime.
```
Comment (btkostner) [inline data/scripts/valadoc.babel.js]: This file should also load pages like the previous script did. When a user clicks a link in the same domain we fetch the HTML of that link, replacing the current content area with the new content area, but not touching the sidebar. This has the benefit of being faster, and not changing the sidebar when searching.

File: data/scripts/valadoc.babel.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+/*
+* Makes a request to the server to get search results
+*/
+function search (query) {
+  const postData = new FormData()
+  postData.append('query', query)
+  postData.append('curpkg', curpkg)
+
+  return fetch('/search.php', {
```
Comment (benwaffle) [inline data/scripts/valadoc.babel.js]: we can polyfill

File: data/scripts/valadoc.babel.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  if (searchDelay) {
+    clearTimeout(searchDelay) // reset the delay
+  }
+
+  if (html.searchField.value != null && html.searchField.value !== '') {
+    // if search isn't empty, we display results after `config.searchDelay` milliseconds
+    searchDelay = setTimeout(() => {
+        search(html.searchField.value).then(res => {
```
Comment (benwaffle) [inline data/scripts/valadoc.babel.js]: indentation

File: data/scripts/valadoc.babel.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+function setupTooltip (link) {
+  link.addEventListener('mouseenter', () => {
+    if (link.getAttribute('data-init') != 'yes') {
+      // fullname = path without the / at the beggining and the .htm(l)
+      const fullname = link.pathname.substring(1).replace(/\.html?$/, '')
+      link.setAttribute('data-init', 'yes')
+      fetch(`/tooltip.php?fullname=${encodeURIComponent(fullname)}`, {
+        method: 'POST'
+      })
+      .then(res => {
```
Comment (benwaffle) [inline data/scripts/valadoc.babel.js]: move to previous line

File: data/scripts/valadoc.babel.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  html.searchFocused = null // The search result that is currently focused
+
+  // Init search
+  html.searchBox.style.display = 'inline-block' // display it (we do it here, so the user without javascript won't see a non-working search box)
+
+  // We run a search when the value changes, after a given delay
+  html.searchField.addEventListener('keyup', evt => {
+    // only if the pressed key isn't up/down arrow, because we use them to select next/previous search result and not to trigger search
+    if (evt.keyCode !== 40 && evt.keyCode !== 38) {
+      updateSearch ()
```
Comment (benwaffle) [inline data/scripts/valadoc.babel.js]: remove space after function name

File: data/scripts/valadoc.babel.js (hunk 3 above)
Comment (Bat41) [inline data/scripts/valadoc.babel.js]: `FormData` is quite well supported too: http://caniuse.com/#search=FormData

**Comment by Bat41** [comment]
Thanks for the review, I'll fix all these problems (and add a polyfill for fetch). I think I will also setup eslint to lint JS.

**Comment by btkostner** [comment]
I vote we should use Standard to lint our JavaScript.

**Comment by Bat41** [comment]
I think it should be OK now. :slightly_smiling_face:

**Comment by btkostner** [comment]
Opps, sorry for the huge delay. It seems GitHub didn't notify my about the changes. I'll test these out soon!

**Comment by btkostner** [comment]
Code looks good and works for me. :+1:

I would like someone else to double check though. Maybe @benwaffle?

**Comment by benwaffle** [comment]
When clicking on anything I get a flash of 'Network Error', the page doesn't load via JS, so the browser treats it as a link and (successfully) does a full load of the page

File: package.json [hunk 7]
```diff
@@ -8,17 +8,25 @@
   "scripts": {
     "test": "stylelint data/styles/*.css",
     "build": "gulp",
-    "start": "php -S localhost:7777 -t ./valadoc.org ./valadoc.org/router.php"
+    "start": "php -S localhost:7777 -t ./valadoc.org ./valadoc.org/router.php",
+    "lint": "./node_modules/.bin/eslint data/scripts/*.js"
```
Comment (benwaffle) [inline package.json]: just write `eslint`, `node_modules/.bin` is imported into the path

File: data/scripts/valadoc.js [hunk 8]
```diff
@@ -1,355 +1,217 @@
...
   }
 }
 
-$(document).ready(function () {
-  $(window).bind('popstate', function (event) {
-    if (window.location.pathname === '' || window.location.pathname === '/') {
-      load_content('index.htm', false)
-    } else {
-      load_content(window.location.pathname, false)
+document.addEventListener('popstate', () => loadPage(window.location))
```
Comment (benwaffle) [inline data/scripts/valadoc.js]: since `loadPage` returns `evt => ...`, this shouldn't be an arrow function

File: data/scripts/valadoc.js [hunk 9]
```diff
@@ -1,355 +1,217 @@
...
-  var seltop = sel.position().top
-  var selbottom = seltop + sel.outerHeight()
-  var box = $('#search-results')
-  var boxtop = box.position().top
-  var boxbottom = boxtop + box.outerHeight()
-  if (seltop < boxtop) {
-    box.scrollTop(box.scrollTop() - (boxtop - seltop))
-  } else if (selbottom > boxbottom) {
-    box.scrollTop(box.scrollTop() + (selbottom - boxbottom))
+      evt.preventDefault()
```
Comment (benwaffle) [inline data/scripts/valadoc.js]: why is this here? Shouldn't it be inside the fetch() callback?

File: data/scripts/valadoc.js [hunk 10]
```diff
@@ -1,355 +1,217 @@
...
-  }
+  link.addEventListener('mouseenter', () => {
+    if (link.getAttribute('data-init') !== 'yes') {
+      // fullname = path without the / at the beggining and the .htm(l)
+      const fullname = link.pathname.substring(1).replace(/\.html?$/, '')
+      link.setAttribute('data-init', 'yes')
+      fetch(`/tooltip.php?fullname=${encodeURIComponent(fullname)}`, {
+        method: 'POST'
+      }).then(res => {
+        return res.text()
```
Comment (benwaffle) [inline data/scripts/valadoc.js]: make one line

**Comment by Bat41** [comment]
Thanks for the review @benwaffle, I fixed all the problems you mentioned. Travis seems to have a problem, but the tests should pass, I don't think my changes can break anything.

**Comment by benwaffle** [comment]
@Bat41 ^

File: data/scripts/valadoc.js [hunk 11]
```diff
@@ -73,7 +73,7 @@ function setupLink (link) {
   link.addEventListener('click', loadPage(link))
 }
 
-function loadPage (link) {
+function loadPage (link, poped) {
```
Comment (benwaffle) [inline data/scripts/valadoc.js]: popped

File: data/scripts/valadoc.js [hunk 12]
```diff
@@ -82,7 +82,9 @@ function loadPage (link) {
 
     fetch(pageUrl).then(res => res.text()).then(page => {
       html.content.innerHTML = page
-      history.pushState(null, title, pageUrl.replace('.content.tpl', ''))
+      if (!poped) { // only add this page to the history again if we didn't visited it juste before, else we won't be able to go back anymore
```
Comment (benwaffle) [inline data/scripts/valadoc.js]: juste -> just
